File: MMMLPDS/VectorAdd/vec_add_mpi.py
from mpi4py import MPI
import numpy as np
import time

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

N = 10_000_000
local_n = N // size

if rank == 0:
    A = np.arange(N, dtype=np.float64)
    B = np.ones(N, dtype=np.float64)
else:
    A = None; B = None

localA = np.empty(local_n, dtype=np.float64)
localB = np.empty(local_n, dtype=np.float64)

comm.Barrier()  # sync before timing
start = MPI.Wtime()

comm.Scatter([A, MPI.DOUBLE], [localA, MPI.DOUBLE], root=0)
comm.Scatter([B, MPI.DOUBLE], [localB, MPI.DOUBLE], root=0)

localC = localA + localB

if rank == 0:
    C = np.empty(N, dtype=np.float64)
else:
    C = None

comm.Gather([localC, MPI.DOUBLE], [C, MPI.DOUBLE], root=0)

comm.Barrier()
end = MPI.Wtime()

if rank == 0:
    print("C[0], C[-1] =", C[0], C[-1])
    print("MPI runtime with {} processes: {:.6f} seconds".format(size, end - start))

File: MMMLPDS/VectorAdd/vec_add_mpi.jl
# vec_add_mpi.jl
using MPI
MPI.Init()
comm = MPI.COMM_WORLD
rank = MPI.Comm_rank(comm)
size = MPI.Comm_size(comm)

N = 1000000
local_n = div(N, size)

if rank==0
    A = collect(0.0:(N-1))
    B = ones(Float64, N)
else
    A = nothing; B = nothing
end

localA = zeros(Float64, local_n)
localB = zeros(Float64, local_n)

MPI.Scatter!(A, localA, 0, comm)
MPI.Scatter!(B, localB, 0, comm)

localC = localA .+ localB

if rank==0
    C = zeros(Float64, N)
else
    C = nothing
end

MPI.Gather!(localC, C, 0, comm)

if rank==0
    println(C[1], " ", C[end])
end

MPI.Finalize()